From 80ed9c51c753e770da565f69f8d6369248324b73 Mon Sep 17 00:00:00 2001
From: "Steven Rostedt (VMware)" <rostedt@goodmis.org>
Date: Mon, 1 May 2017 15:45:23 -0400
Subject: [PATCH] trace-cmd record: Fix filtering when using set_event_pid

BZ: BZ1447927

When set_event_pid exists, trace-cmd uses it for filtering tasks. A commit
that updates the pid filtering was not updated to handle excludes of the
tracing, and they were added too, causing more to be filtered than what
should have been.

Fixes: a5923c5c9f81 ("trace-cmd: Filter out specific pids")
Signed-off-by: Steven Rostedt (VMware) <rostedt@goodmis.org>
Signed-off-by: John Kacur <jkacur@redhat.com>
---
 trace-record.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/trace-record.c b/trace-record.c
index bd60892ae621..1b55043e2842 100644
--- a/trace-record.c
+++ b/trace-record.c
@@ -2042,11 +2042,16 @@ static void update_pid_filters(struct buffer_instance *instance)
 	str = filter;
 
 	for (p = filter_pids; p; p = p->next) {
+		if (p->exclude)
+			continue;
 		len = sprintf(str, "%d ", p->pid);
 		str += len;
 	}
 
-	len = len_filter_pids + nr_filter_pids;
+	if (filter == str)
+		goto out;
+
+	len = str - filter;
 	str = filter;
 	do {
 		ret = write(fd, str, len);
@@ -2056,6 +2061,7 @@ static void update_pid_filters(struct buffer_instance *instance)
 		len -= ret;
 	} while (ret >= 0 && len);
 
+ out:
 	close(fd);
 }
 
-- 
2.5.5

